### PR TITLE
W-12597177 - Add support for multipart/form-data requests with connec…

### PR DESCRIPTION
**Pull request by andrewslater**
…tion.js

This change introduces new options to `DmlOptions` where individual fields can be marked to be sent as multipart/form-data elements. This is useful for sending requests that are > 52mb (limit when using application/json content-type). See tests for details on usage. Also note it is unreasonable to expect this code to break any existing consumers of this library because all change in behavior is gated on the use of the new `multipartFileFields` option.

### CODE REVIEW

File: test/tooling.test.ts [hunk 1]
```diff
@@ -90,6 +91,57 @@ it('should get completions and return completions', async () => {
...
+        filename: 'test.zip',
+      },
+    },
+  };
+  const record = await conn.tooling.create('StaticResource', request, options);
+  assert.ok(record.success);
+});
+
+it('can create static resource using multipart/form-data content type with a base64 encoded string', async () => {
+  const request = createStaticResourceRequest(
```
Comment (ahmedrana2022) [inline test/tooling.test.ts:119]: question: wondering why we would need to base-64 encode the file, wouldn't the caller just pass the zip file (like in the previous test) ?

File: src/connection.ts [hunk 2]
```diff
@@ -969,13 +970,41 @@ export class Connection<S extends Schema = Schema> extends EventEmitter {
...
+          form.append(
+            fieldName,
+            Buffer.from(rec[fieldName], 'base64'),
+            fileDetails,
+          );
+          delete rec[fieldName];
+        },
+      );
+      // Serialize the remaining fields as JSON
+      form.append(type, JSON.stringify(rec), {
```
Comment (ganeshnt) [inline src/connection.ts:990]: With my limited knowledge on TS... Logic is in this block of execution only when the conditional at #975 is satisfied, which establish that we are in this block to process multi-part requested fields.  So the the loop starting on #979 iterate to process each field that needs multi-part function.  If my assessment of this logic is correct, then i do not understand what is left to process on line #990?

Comment (ganeshnt) [reply src/connection.ts:990]: if there were fields that need JSON processing, shouldn't this branch into the "else" clause below?

File: test/data/test.zip
Comment (ganeshnt) [inline test/data/test.zip:1]: what do we have in the zip? and Git shows +348 bytes only!

**Review by ganeshnt — APPROVED** [review]
with my limited knowledge on TS and this area of code, the review looks okay to me and i do not see any reason for potential regressions to other areas considering this change is way down the stack

File: src/connection.ts (hunk 2 above)
Comment (andrewslater) [reply src/connection.ts:990]: What remains is any fields that were not referenced in `multipartFileFields`. The key is the `delete rec[fieldName]` on line 986. Consider a hypothetical `rec` for an entity named `MyEntity" is passed in:
```
{
  Name: "MyRecordName",
  Cacheable: True,
  MyFileField: "<base64 encoded string or buffer>"
}
```

with `options.multipartFileFields`:
```
{
  MyFileFIeld: {
    contentType: "image/png",
    fileName: "company-logo.png"
  }
}
```

After the loop starting on line 979 completes, we have a `FormData` object that contains (when serialized):
```
----------------------------248819749774746952254207
Content-Disposition: form-data; name="MyFileField"; filename="company-logo.png" 
Content-Type: image/png
<base64 encoded file content>
```

We still need to add the remaining fields so line 990 serializes the next "part" of our request which results in:
```
----------------------------248819749774746952254207
Content-Disposition: form-data; name="MyEntity"
Content-Type: application/json
 
{"Name": "MyRecordName", "Cacheable": true}
```

Now all parts will be sent, with the appropriate content type.

File: test/tooling.test.ts (hunk 1 above)
Comment (andrewslater) [reply test/tooling.test.ts:119]: While iterating on this I had some (incorrect) changes to  `request-helper.ts` that resulted in passing a Buffer to fail but passing a string to work, so I wanted to have tests for both scenarios.

Comment (andrewslater) [reply test/tooling.test.ts:119]: It passes a base64 encoded string that represents a zip file. The content type for a multipart "part" is not the same as the "Content-Type" HTTP header. Each "part" has to specify what its content type is. This is the equivalent of the curl command option you referenced in your research:

```
-F 'versionInfo=@package-version-info.zip;type=application/zip'
```

For binary content types (as opposed to, say, the json "part" we send for the non-file fields) you specify a content type for the _"part"_ and base64 encode the contents. In this case the `application/zip` content type for the `Body` part is used to tell the server what type of data will be present after decoding the base64 encoded string.

I tested that the StaticResource is saved and is identical when retrieved in both cases.

**Review by ganeshnt — APPROVED** [review]

File: test/tooling.test.ts (hunk 1 above)
Comment (ahmedrana2022) [reply test/tooling.test.ts:119]: Got it, thanks for clarifying that.

**Review by ahmedrana2022 — APPROVED** [review]

**Review by cristiand391 — APPROVED** [review]

**Pull request merged**